src/tests/serde.rs

```diff
@@ -594,3 +600,188 @@ fn test_serialize_deserialize_unsigned_numbers() {
+        pub date: chrono::DateTime<chrono::Utc>,
+    }
+
+    let date = r#"
+    {
+        "date": {
+                "$date": {
+                    "$numberLong": "1591700287095"
+                }
+        }
+    }"#;
+    let json: Value = serde_json::from_str(&date).unwrap();
+    let b: B = serde_json::from_value(json).unwrap();
+    let expected: chrono::DateTime<chrono::Utc> =
+        chrono::DateTime::from_str("2020-06-09 10:58:07.095 UTC").unwrap();
+    assert_eq!(b.date, expected);
+
+    #[derive(Deserialize, Serialize)]
+    struct C {
+        #[serde(
+            serialize_with = "serde_helpers::serialize_iso_string_as_datetime",
+            deserialize_with = "serde_helpers::deserialize_iso_string_from_datetime"
+        )]
+        pub date: String,
+    }
+
+    let date = "2020-06-09 10:58:07.095 UTC";
+    let c = C {
+        date: date.to_string(),
+    };
+    let doc = to_document(&c).unwrap();
+    assert!(doc.get_datetime("date").is_ok());
+    let c: C = from_document(doc).unwrap();
+    assert_eq!(c.date.as_str(), date);
+}
+
+#[test]
+fn test_oid_helpers() {
+    let _guard = LOCK.run_concurrently();
+
+    #[derive(Deserialize, Serialize)]
+    struct A {
+        #[serde(deserialize_with = "serde_helpers::deserialize_object_id_from_ext_json")]
+        oid: ObjectId,
+    }
+
+    let oid = ObjectId::new();
+    let a = A { oid: oid.clone() };
+    let bson = to_bson(&a).unwrap();
+    let value: Value = bson.into();
+    let a: A = serde_json::from_value(value).unwrap();
+    assert_eq!(a.oid, oid);
+
+    #[derive(Serialize)]
+    struct B {
+        #[serde(serialize_with = "serde_helpers::serialize_hex_string_as_object_id")]
+        oid: String,
+    }
+
+    let oid = ObjectId::new();
+    let b = B {
+        oid: oid.to_string(),
+    };
+    let doc = to_document(&b).unwrap();
+    assert_eq!(doc.get_object_id("oid").unwrap(), &oid);
+}
+
+#[test]
+fn test_uuid_helpers() {
+    let _guard = LOCK.run_concurrently();
+
+    #[derive(Serialize, Deserialize)]
+    struct A {
+        #[serde(
+            serialize_with = "serde_helpers::serialize_uuid_as_binary",
+            deserialize_with = "serde_helpers::deserialize_uuid_from_binary"
+        )]
+        uuid: Uuid,
+    }
+
+    let uuid = Uuid::parse_str("936DA01F9ABD4d9d80C702AF85C822A8").unwrap();
+    let a = A { uuid };
+    let doc = to_document(&a).unwrap();
+    match doc.get("uuid").unwrap() {
+        Bson::Binary(bin) => {
+            assert_eq!(bin.subtype, BinarySubtype::Uuid);
+            assert_eq!(bin.bytes, uuid.as_bytes());
+        }
+        _ => panic!("expected Bson::Binary"),
+    }
+    let a: A = from_document(doc).unwrap();
+    assert_eq!(a.uuid, uuid);
+}
+
+#[test]
+fn test_timestamp_helpers() {
+    let _guard = LOCK.run_concurrently();
+
+    #[derive(Deserialize, Serialize)]
+    struct A {
+        #[serde(
+            serialize_with = "serde_helpers::serialize_u64_as_timestamp",
```

Comment:
Ditto above about considering a module for allowing roundtripping with a single attribute.

Reply:
Not sure I understand the differentiation between `u64`/`Timestamp` and `Uuid`/`Binary` here -- wouldn't it be possible to implement a simple helper there as well? I think the benefit here is similar to that of the previous serde work to accept user types into crud methods in that it allows users not to do any conversions/not need to think about `Timestamp` (besides adding the attribute) and instead working directly with `u64`.

Reply:
I think one issue here is that timestamps, when represented as integers, are typically the number of seconds since the epoch, but the BSON timestamp type supports sub-second accuracy by means of the ordinal. This means using the `deserialize_u64_from_timestamp` is lossy for any timestamp with an ordinal assuming we're using the unix timestamp as the format of our `u64`. We could opt to use the `u64` as the storage of the timestamp as it's used in BSON, but I think that would be confusing to users who are likely expecting the number to be a unix timestamp. 

Given these concerns, I don't think we necessarily want to encourage users to use `u64` in place of `Timestamp` when deserializing from BSON or extended JSON. The serialization case seems ok, and both serialization to / deserialization from `u64` for `Timestamp` should be fine as well, since it is the more precise type. In the UUID case, the conversion is lossless, so it's okay to use the `Uuid` type as a drop-in replacement for BSON binary. If a `u64` (as a unix timestamp) could be losslessly converted to/from BSON timestamp, then I'd agree that there is no meaningful difference though.

Reply:
I believe `u64 -> Timestamp` can also be lossy, since the seconds since the Unix epoch is stored in a `u32` in `Timestamp` (so anything greater than `u32::MAX` would be truncated). We do have [one user](https://github.com/mongodb/bson-rust/issues/122) issue asking for this functionality though, so it might be best to assume the `u64` is seconds from the Unix epoch and error out in the case of a lossy conversion (and make sure that behavior is well-documented). Alternatively, we could include sets of functions for both seconds from the Unix epoch and the BSON storage including an increment, but I think it would be best to hold off on the latter unless someone asks for it.

Reply:
Oh yeah that's a good point. It seems like we actually just want conversions between `u32` and what not then, if we're going to be erroring on anything greater anyways.

And yeah I agree. I doubt anyone is going to ask for having the BSON storage as an actual integer.